packages/material-ui/src/styles/createTypography.js

```diff
@@ -15,118 +26,229 @@ export default function createTypography(palette, typography) {
+        textTransform: 'uppercase',
+      },
+    },
+    ignoreDeprecationWarnings = process.env.MUI_SUPPRESS_DEPRECATION_WARNINGS,
+    useNextVariants = false,
     // Apply the CSS properties to all the variants.
     allVariants,
     ...other
   } = typeof typography === 'function' ? typography(palette) : typography;
 
+  warning(
+    !Object.keys(other).some(variant => deprecatedVariants.includes(variant)),
+    'Deprecation Warning: Material-UI: You are passing a deprecated variant to ' +
+      `createTypography. ${migrationGuideMessage}`,
+  );
+
+  warning(
+    useNextVariants || !Object.keys(other).some(variant => restyledVariants.includes(variant)),
+    'Deprecation Warning: Material-UI: You are passing a variant to createTypography ' +
+      'that will be restyled in the next major release, without indicating that you ' +
+      `are using typography v2 (set \`useNextVariants\` to true. ${migrationGuideMessage}`,
+  );
+
   const coef = fontSize / 14;
-  function pxToRem(value) {
-    return `${(value / htmlFontSize) * coef}rem`;
-  }
+
+  const letterSpacingToEm = (tracking, spSize) => {
+    return `${(tracking / spSize) * coef}em`;
+  };
+
+  const pxToRem = size => {
+    return `${(size / htmlFontSize) * coef}rem`;
+  };
+
+  const getVariant = (variant, localUseNextVariants) => {
+    warning(
+      ignoreDeprecationWarnings || !deprecatedVariants.includes(variant),
+      'Deprecation Warning: Material-UI: You are using the deprecated typography variant ' +
+        `${variant} that will be removed in the next major release. ${migrationGuideMessage}`,
+    );
+
+    const nextVariant = nextVariantMapping(variant);
+
+    // complete v2 switch
+    if (useNextVariants) {
+      return nextVariant;
+    }
+
+    const isRestyledVariant = restyledVariants.includes(variant);
+    // global v1, local v2
+    if (localUseNextVariants) {
+      if (isRestyledVariant) {
+        return `${nextVariant}Next`;
+      }
+      return nextVariant;
+    }
+
+    // v1 => restyle warnings
+    warning(
+      ignoreDeprecationWarnings || !isRestyledVariant,
+      'Deprecation Warning: Material-UI: You are using the typography variant ' +
+        `${variant} which will be restyled in the next major release. ${migrationGuideMessage}`,
+    );
+
+    return variant;
+  };
+
+  const utils = { getVariant, letterSpacingToEm, pxToRem };
+
+  const propertiesForCategory = (weight, size, casing, letterSpacing) => {
+    return {
+      color: palette.text.primary,
+      fontFamily,
+      fontSize: pxToRem(size),
+      fontWeight: weight,
+      letterSpacing: letterSpacingToEm(letterSpacing, size),
+      ...casing,
+      ...allVariants,
+    };
+  };
+
+  /* eslint-disable key-spacing, no-multi-spaces */
+  // prettier-ignore
+  const nextVariants = {
+    headline1:   propertiesForCategory(fontWeightLight,   96, caseSentence, -1.5),
+    headline2:   propertiesForCategory(fontWeightLight,   60, caseSentence, -0.5),
```

Comment:
Having `caseSentence` as a default behavior is opinionated. I think that we will be better off not having it by default. Very few, to no CSS libraries comes with this style as a default. I don't think that people expect this behavior. Looking at the source code of our application at work, we have only one use case for `'&:first-letter': { textTransform: 'uppercase' },` while many for disabling it. So, what do you think of removing this default behavior and adding a property to enable it? (But I'm not even sure we need a property for that).

Reply:
The specification defines it that way.

Reply:
This is already something that can be disabled with:
```js
createTheme({
  typography: {
    caseSentence: {}
  }
})
```
Maybe add a documentation for the `createTypography` part of `createTheme`.

Reply:
And yet, MCW isn't implementing it either.

Reply:
Case made. Use `inherit` instead or just leave it blank?

Reply:
What do you mean by using inherit?